// monofile-utilities/lib/indent-builder.mjs
/*!
 *
 * Copyright 2018 - acrazing
 *
 * @author acrazing joking.young@gmail.com
 * @since 2018-01-26 23:34:33
 * @version 1.0.0
 * @desc indent-builder.ts
 */
import { __values } from "tslib";
var IndentBuilder = /** @class */ (function () {
    function IndentBuilder(element) {
        if (element === void 0) { element = '  '; }
        this.element = element;
        this.content = '';
        this.prefix = '';
    }
    IndentBuilder.prototype.indent = function () {
        this.prefix += this.element;
        return this;
    };
    IndentBuilder.prototype.push = function () {
        var e_1, _a;
        var lines = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            lines[_i] = arguments[_i];
        }
        try {
            for (var lines_1 = __values(lines), lines_1_1 = lines_1.next(); !lines_1_1.done; lines_1_1 = lines_1.next()) {
                var line = lines_1_1.value;
                this.content += this.prefix + line + '\n';
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (lines_1_1 && !lines_1_1.done && (_a = lines_1.return)) _a.call(lines_1);
            }
            finally { if (e_1) throw e_1.error; }
        }
        return this;
    };
    IndentBuilder.prototype.undent = function () {
        this.prefix = this.prefix.substr(0, this.prefix.length
            - this.element.length);
        return this;
    };
    return IndentBuilder;
}());
export { IndentBuilder };
