// monofile-utilities/lib/is.js
"use strict";
/*!
 * Copyright 2018 acrazing <joking.young@gmail.com>. All rights reserved.
 * @since 2018-07-30 19:05:19
 */
Object.defineProperty(exports, "__esModule", { value: true });
const falsy = (value) => false;
exports.isObject = (value) => typeof value === 'object' && value !== null;
exports.isArray = (value) => value instanceof Array;
exports.isDate = (value) => value instanceof Date;
exports.isError = (value) => value instanceof Error ||
    (exports.isObject(value) && 'name' in value && 'message' in value);
exports.isMap = typeof Map === 'function'
    ? (value) => value instanceof Map
    : falsy;
exports.isBlob = typeof Blob === 'function'
    ? (input) => input instanceof Blob
    : (input) => false;
exports.isFile = typeof File === 'function'
    ? (input) => input instanceof File
    : (input) => false;
exports.isBufferSource = typeof ArrayBuffer === 'function'
    ? (input) => input instanceof ArrayBuffer || ArrayBuffer.isView(input)
    : (input) => false;
exports.isArrayBufferView = typeof ArrayBuffer === 'function'
    ? ArrayBuffer.isView
    : (input) => false;
exports.isFormData = typeof FormData === 'function'
    ? (input) => input instanceof FormData
    : (input) => false;
exports.isURLSearchParams = typeof URLSearchParams === 'function'
    ? (input) => input instanceof URLSearchParams
    : (input) => false;
exports.isReadableStream = typeof ReadableStream === 'function'
    ? (input) => input instanceof ReadableStream
    : (input) => false;
exports.isBodyInit = (input) => typeof input === 'string' ||
    exports.isBlob(input) ||
    exports.isBufferSource(input) ||
    exports.isFormData(input) ||
    exports.isURLSearchParams(input) ||
    exports.isReadableStream(input);
