// monofile-utilities/lib/is.mjs
/*!
 * Copyright 2018 acrazing <joking.young@gmail.com>. All rights reserved.
 * @since 2018-07-30 19:05:19
 */
var falsy = function (value) { return false; };
export var isObject = function (value) {
    return typeof value === 'object' && value !== null;
};
export var isArray = function (value) {
    return value instanceof Array;
};
export var isDate = function (value) { return value instanceof Date; };
export var isError = function (value) {
    return value instanceof Error ||
        (isObject(value) && 'name' in value && 'message' in value);
};
export var isMap = typeof Map === 'function'
    ? function (value) { return value instanceof Map; }
    : falsy;
export var isBlob = typeof Blob === 'function'
    ? function (input) { return input instanceof Blob; }
    : function (input) { return false; };
export var isFile = typeof File === 'function'
    ? function (input) { return input instanceof File; }
    : function (input) { return false; };
export var isBufferSource = typeof ArrayBuffer === 'function'
    ? function (input) {
        return input instanceof ArrayBuffer || ArrayBuffer.isView(input);
    }
    : function (input) { return false; };
export var isArrayBufferView = typeof ArrayBuffer === 'function'
    ? ArrayBuffer.isView
    : function (input) { return false; };
export var isFormData = typeof FormData === 'function'
    ? function (input) { return input instanceof FormData; }
    : function (input) { return false; };
export var isURLSearchParams = typeof URLSearchParams === 'function'
    ? function (input) {
        return input instanceof URLSearchParams;
    }
    : function (input) { return false; };
export var isReadableStream = typeof ReadableStream === 'function'
    ? function (input) {
        return input instanceof ReadableStream;
    }
    : function (input) { return false; };
export var isBodyInit = function (input) {
    return typeof input === 'string' ||
        isBlob(input) ||
        isBufferSource(input) ||
        isFormData(input) ||
        isURLSearchParams(input) ||
        isReadableStream(input);
};
