// monofile-utilities/lib/is.spec.js
"use strict";
/*!
 * Copyright 2019 acrazing <joking.young@gmail.com>. All rights reserved.
 * @since 2019-12-05 16:38:39
 */
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const _is = tslib_1.__importStar(require("./is"));
const is = _is;
describe('is', () => {
    it('should check as expected', () => {
        const values = {
            Date: () => new Date(),
            Object: () => ({}),
            Array: () => [[]],
            Blob: () => new Blob(),
            File: () => new File([], ''),
            BufferSource: () => [
                new ArrayBuffer(1),
                Uint32Array.of(1),
                new DataView(new ArrayBuffer(1)),
            ],
            ArrayBufferView: () => Uint8Array.of(1),
            FormData: () => new FormData(),
            URLSearchParams: () => new URLSearchParams(),
            ReadableStream: () => new ReadableStream(),
            BodyInit: () => '',
        };
        Object.keys(values).forEach((kind) => {
            let items;
            const method = is['is' + kind];
            try {
                items = values[kind]();
                items = Array.isArray(items) ? items : [items];
            }
            catch (_a) {
                expect([kind, method(global)]).toEqual([kind, false]);
                return;
            }
            items.forEach((item) => {
                expect([kind, item, method(item)]).toEqual([kind, item, true]);
            });
        });
    });
});
