// monofile-utilities/lib/request-guid.express.spec.js
"use strict";
/*!
 * Copyright 2019 acrazing <joking.young@gmail.com>. All rights reserved.
 * @since 2019-01-25 20:30:49
 */
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const express_1 = tslib_1.__importDefault(require("express"));
const node_fetch_1 = tslib_1.__importDefault(require("node-fetch"));
const request_guid_1 = require("./request-guid");
const request_guid_express_1 = require("./request-guid.express");
describe('express middleware', function () {
    it('should set header correctly', () => tslib_1.__awaiter(this, void 0, void 0, function* () {
        const header = 'x-request-id-2';
        const from = 'AAAAAAAA-AAAAAAAA-AAAAAAAA-AAAAAAAA';
        return new Promise((resolve, reject) => {
            const app = express_1.default();
            app.use(request_guid_express_1.requestGuid({
                secret: '13456789abcdef0',
                trustHeader: true,
                name: 'id',
                header: header,
            }));
            app.get('/', (req, res) => {
                expect(req.id).toMatch(request_guid_1.GUID_RE);
                res.end('');
            });
            const srv = app.listen(2345, '127.0.0.1', () => {
                new Promise((resolve) => setTimeout(resolve, 0))
                    .then(() => {
                    return node_fetch_1.default('http://127.0.0.1:2345/').then((r) => {
                        expect(r.headers.get(header)).toMatch(request_guid_1.GUID_RE);
                        expect(r.headers.get(header)).not.toBe(from);
                    });
                })
                    .then(() => {
                    return node_fetch_1.default('http://127.0.0.1:2345/', {
                        headers: { [header]: from },
                    }).then((r) => {
                        expect(r.headers.get(header)).toBe(from);
                    });
                })
                    .then(resolve, reject)
                    .then(() => {
                    srv.close();
                });
            });
        });
    }));
});
