// monofile-utilities/lib/storage.js
"use strict";
/*!
 *
 * Copyright 2017 - acrazing
 *
 * @author acrazing joking.young@gmail.com
 * @since 2017-12-15 12:29:38
 * @version 1.0.0
 * @desc storage.ts
 */
Object.defineProperty(exports, "__esModule", { value: true });
class CookieStorage {
    get(name, defaults) {
        const key = encodeURIComponent(name);
        const matches = document.cookie.match(new RegExp(`(?:^|; )${key.replace(/[.*()]/g, '\\$&')}=([^;]*)`));
        const result = matches ? decodeURIComponent(matches[1]) : defaults;
        try {
            return JSON.parse(result);
        }
        catch (e) {
            return result;
        }
    }
    set(name, value, opts = {}) {
        typeof value === 'string' || (value = JSON.stringify(value));
        let cookie = encodeURIComponent(name) + '=' + encodeURIComponent(value);
        opts.maxAge && (cookie += '; Max-Age=' + opts.maxAge);
        typeof opts.expires === 'number' && (opts.expires =
            new Date(+new Date + opts.expires * 1e3));
        opts.expires instanceof Date && (cookie +=
            `; Expires=${opts.expires.toUTCString()}`);
        opts.path && (cookie += '; Path=' + opts.path);
        opts.domain && (cookie += '; Domain=' + opts.domain);
        opts.secure && (cookie += '; Secure');
        opts.httpOnly && (cookie += '; HttpOnly');
        document.cookie = cookie;
        return this;
    }
    remove(name) {
        this.set(name, '', { maxAge: 0 });
        return this;
    }
}
exports.CookieStorage = CookieStorage;
class LocalStorage {
    get(name, defaults) {
        let result = localStorage.getItem(name);
        result === null && (result = void 0);
        try {
            const data = JSON.parse(result);
            if (!data || !data.hasOwnProperty('__')) {
                // could be parse, but not internal value
                // return the value
                return data === void 0 ? defaults : data;
            }
            if (data.expires && data.expires < +new Date) {
                localStorage.removeItem(name);
                return defaults;
            }
            return data.data;
        }
        catch (e) {
            return result === void 0 ? defaults : result;
        }
    }
    set(name, value, opts = {}) {
        const data = {
            __: true,
            data: value,
        };
        opts.expires && (data.expires = opts.expires instanceof Date
            ? +opts.expires : (+new Date + opts.expires * 1e3));
        localStorage.removeItem(name);
        localStorage.setItem(name, JSON.stringify(data));
        return this;
    }
    remove(name) {
        localStorage.removeItem(name);
        return this;
    }
}
exports.LocalStorage = LocalStorage;
class NoopStorage {
    get(_name, defaults) {
        return defaults;
    }
    set() {
        return this;
    }
    remove() {
        return this;
    }
}
exports.NoopStorage = NoopStorage;
exports.hasLocal = typeof window !== 'undefined'
    && !!window.localStorage
    && !!window.localStorage.getItem
    && !!window.localStorage.setItem
    && !!window.localStorage.removeItem;
const noop = new NoopStorage();
exports.cookie = new CookieStorage();
exports.local = exports.hasLocal ? new LocalStorage() : noop;
exports.storage = exports.hasLocal ? exports.local : exports.cookie;
