// monofile-utilities/lib/Unique.mjs
/*!
 * Copyright 2019 acrazing <joking.young@gmail.com>. All rights reserved.
 * @since 2019-02-18 14:48:28
 */
var Unique = /** @class */ (function () {
    function Unique(minId, maxId) {
        this.minId = minId;
        this.free = maxId - minId;
        this.pool = new Uint32Array(this.free);
        this.pos = new Uint32Array(this.free);
        for (var i = 0; i < this.free; i++) {
            this.pool[i] = this.pos[i] = i;
        }
    }
    Unique.prototype.rent = function () {
        if (this.free === 0) {
            return null;
        }
        var index = Math.floor(Math.random() * this.free);
        var value = this.pool[index];
        var lastIndex = this.free - 1;
        var lastValue = this.pool[lastIndex];
        this.pool[lastIndex] = value;
        this.pool[index] = lastValue;
        this.pos[value] = lastIndex;
        this.pos[lastValue] = index;
        this.free -= 1;
        return value + this.minId;
    };
    Unique.prototype.return = function (id) {
        id = id - this.minId;
        var index = this.pos[id];
        if (index === void 0 || index < this.free) {
            return;
        }
        var firstValue = this.pool[this.free];
        this.pool[this.free] = id;
        this.pool[index] = firstValue;
        this.pos[firstValue] = index;
        this.pos[id] = this.free;
        this.free += 1;
    };
    return Unique;
}());
export { Unique };
