// monofile-utilities/lib/validate.number.js
"use strict";
/*!
 * Copyright 2019 acrazing <joking.young@gmail.com>. All rights reserved.
 * @since 2019-12-06 22:08:17
 */
Object.defineProperty(exports, "__esModule", { value: true });
const consts_1 = require("./consts");
function only(value, accepts) {
    return accepts.indexOf(value) > -1;
}
exports.only = only;
function gt(value, than) {
    return value > than;
}
exports.gt = gt;
function lt(value, than) {
    return value < than;
}
exports.lt = lt;
function gte(value, than) {
    return value >= than;
}
exports.gte = gte;
function lte(value, than) {
    return value <= than;
}
exports.lte = lte;
function isInt(value, gte = Number.MIN_SAFE_INTEGER, lte = Number.MAX_SAFE_INTEGER) {
    return value === Math.floor(value) && value >= gte && value <= lte;
}
exports.isInt = isInt;
function isUInt(value) {
    return isInt(value, 0);
}
exports.isUInt = isUInt;
function isPInt(value) {
    return isInt(value, 1);
}
exports.isPInt = isPInt;
function isUInt32(value) {
    return isInt(value, 0, consts_1.UINT32_MAX);
}
exports.isUInt32 = isUInt32;
function isUInt31(value) {
    return isInt(value, 0, consts_1.INT32_MAX);
}
exports.isUInt31 = isUInt31;
function isInt32(value) {
    return isInt(value, consts_1.INT32_MIN, consts_1.INT32_MAX);
}
exports.isInt32 = isInt32;
