// gomoku-terminal/lib/utils/service/api.js
"use strict";
/*!
 * Copyright 2019 acrazing <joking.young@gmail.com>. All rights reserved.
 * @since 2019-04-14 17:22:40
 */
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const axios_1 = tslib_1.__importDefault(require("axios"));
const ServiceError_1 = require("./ServiceError");
const packageJson = require('../../../package.json');
const config = {
    host: '',
    token: '',
};
function API(api) {
    return (data) => axios_1.default.post(`${config.host}/api${api}`, data, {
        validateStatus: () => true,
        headers: {
            Authorization: config.token,
            'User-Agent': `Kcats gomoku-terminal/${packageJson.version}`,
        },
    }).then((r) => {
        if (r.status > 299) {
            throw new ServiceError_1.ServiceError(r.data.code, r.data.message);
        }
        return r.data;
    }, () => {
        throw new ServiceError_1.ServiceError(504, '网络繁忙, 请稍候再试');
    });
}
exports.API = API;
API.config = (options) => {
    Object.assign(config, options);
};
exports.userLogin = API('/user.login');
exports.userGet = API('/user.get');
exports.userRegister = API('/user.register');
exports.gomokuGetAccessToken = API('/gomoku.getAccessToken');
