// gomoku-terminal/lib/types/GomokuService.idl.js
"use strict";
/*!
 * Copyright 2019 acrazing <joking.young@gmail.com>. All rights reserved.
 * @since 2019-01-18 23:32:19
 */
Object.defineProperty(exports, "__esModule", { value: true });
const enum_1 = require("monofile-utilities/lib/enum");
const messages_1 = require("./messages");
const RoomModule_idl_1 = require("./RoomModule.idl");
exports.GomokuOverReason = enum_1.Enum({
    WIN: '游戏获胜',
    STEP_TIMEOUT: '步时超时',
    TOTAL_TIMEOUT: '局时超时',
    GIVE_UP: '认输',
    AGREE_PEACE: '平局',
    BOARD_EXHAUSTED: '平局',
});
exports.GomokuGameMode = enum_1.Enum({
    ONLINE: '在线游戏',
    OFFLINE: '离线游戏',
});
exports.GomokuPrefabGroup = enum_1.Enum({
    NORMAL: '普通场',
    MASTER: '高手场',
    GREAT_MASTER: '大师场',
});
exports.GomokuRole = enum_1.Enum({
    BLACK: '先手',
    WHITE: '后手',
});
exports.GomokuPrefabLabels = messages_1.defineLabels(Object.assign({}, RoomModule_idl_1.RoomPrefabLabels, { mode: '游戏类型', group: '分组', gameTimeout: '游戏超时时间', stepTimeout: '步时超时时间', minEnterScore: '最低入场积分', maxEnterScore: '最高入场积分', status: '状态' }));
exports.GomokuGameLabels = messages_1.defineLabels({
    previous: '上一局连续的游戏ID',
    mode: '类型',
    prefabId: '游戏配置',
    blackSeat: '庄家座位',
    overReason: '结束原因',
    winnerRole: '获胜角色',
    record: '操作记录',
});
exports.GomokuGameUserLabels = messages_1.defineLabels({
    gameId: '游戏 ID',
    userId: '用户 ID',
    seat: '座位',
    role: '角色',
    initialScore: '初始积分',
    score: '得分',
    usedTime: '游戏用时',
});
exports.GomokuGameResultLabels = messages_1.defineLabels(Object.assign({}, exports.GomokuGameLabels, { players: '玩家列表' }));
exports.GomokuUserLabels = messages_1.defineLabels({
    userId: '用户 ID',
    onlineScore: '在线游戏积分',
    onlineCount: '在线游戏局数',
    onlineWinCount: '在线游戏获胜局数',
    onlineBlackCount: '在线游戏先手局数',
    onlineBlackWinCount: '在线游戏先手获胜局数',
    onlineMaxWinningStreak: '最高连胜',
    onlineCurrentWinningStreak: '当前连胜',
    offlineScore: '离线游戏积分',
    offlineCount: '离线游戏局数',
    offlineWinCount: '离线游戏获胜局数',
    offlineBlackCount: '离线游戏先手局数',
    offlineBlackWinCount: '离线游戏先手获胜局数',
});
exports.GomokuLabels = Object.assign({
    ids: 'ID 列表',
    gameId: '游戏 ID',
    index: '操作序号',
    location: '落子位置',
}, exports.GomokuUserLabels, exports.GomokuGameLabels, exports.GomokuGameResultLabels, exports.GomokuGameUserLabels, exports.GomokuOverReason.labels);
