// gomoku-terminal/lib/types/RoomModule.idl.js
"use strict";
/*!
 * Copyright 2019 acrazing <joking.young@gmail.com>. All rights reserved.
 * @since 2019-02-16 17:32:26
 */
Object.defineProperty(exports, "__esModule", { value: true });
const type_tree_1 = require("monofile-utilities/lib/type-tree");
var RoomUserStatus;
(function (RoomUserStatus) {
    RoomUserStatus["Free"] = "Free";
    RoomUserStatus["Pending"] = "Pending";
    RoomUserStatus["Unready"] = "Unready";
    RoomUserStatus["Ready"] = "Ready";
    RoomUserStatus["Playing"] = "Playing";
    RoomUserStatus["View"] = "View";
})(RoomUserStatus = exports.RoomUserStatus || (exports.RoomUserStatus = {}));
exports.RoomLabels = {
    roomId: '房间号',
    seat: '座位号',
    prefabId: '房间类型',
};
var RoomExitReason;
(function (RoomExitReason) {
    RoomExitReason["EnterTimeout"] = "EnterTimeout";
    RoomExitReason["ReadyTimeout"] = "ReadyTimeout";
    RoomExitReason["Disconnect"] = "Disconnect";
    RoomExitReason["UserExit"] = "UserExit";
    RoomExitReason["RoomDestroy"] = "RoomDestroy";
})(RoomExitReason = exports.RoomExitReason || (exports.RoomExitReason = {}));
exports.RoomPrefabLabels = {
    id: '预设 id',
    name: '名称',
    seat: '座位数量',
    autoReady: '自动准备',
    readyTimeout: '准备超时时间',
    allowSitInGame: '游戏中允许坐下',
};
exports.RoomEvents = type_tree_1.createTypeTree('room', {
    leave: '',
    enter: '',
    askReady: '',
    ready: '',
    unready: '',
});
