// gomoku-terminal/lib/types/UserService.idl.js
"use strict";
/*!
 * Copyright 2019 acrazing <joking.young@gmail.com>. All rights reserved.
 * @since 2019-01-18 20:17:23
 *
 * XxxService.idl.ts
 *
 * 1. 声明 service 的入参和返回值类型列表, XxxActionInput, XxxActionOutput
 * 2. 声明所有的方法到 XxxServiceIdl 接口中, action(input: XxxActionInput): XxxActionOutput
 * 3. 声明模块名称: SERVICE_NAME_XXX: Xxx = Xxx
 *
 * 似乎与 grpc 中的 service 的 proto 文件类似.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const messages_1 = require("./messages");
exports.UserLabels = messages_1.defineLabels({
    username: '用户名',
    nickname: '昵称',
    password: '密码',
    mobile: '手机号',
    email: '邮箱',
    isAnonymous: '匿名用户',
});
exports.SessionLabels = messages_1.defineLabels({
    token: 'token',
    userId: '用户 id',
    ip: 'IP',
    userAgent: '浏览器',
    previousSessionId: '上游会话 id',
    referer: '来源页面',
    userVisible: '用户可见',
    expiresAt: '过期时间',
    maxAge: '过期时间',
});
exports.AccessTokenLabels = messages_1.defineLabels({
    token: '凭证',
    userId: '用户ID',
    resource: '允许访问的资源',
    expiresAt: '过期时间',
    maxAge: '有效期',
});
