// pop-cloud-request/src/service/activity.ts

import { generatePaginationQuery, extendQuery } from "../tools";


/**获取活动分页列表 */
export function getActivityList(request, params) {
  const { page, pageSize } = params
  return generatePaginationQuery({
    client: extendQuery(request, {}, '/v1/activity'),
    name: 'activities',
    where: {},
    order: {},
    page,
    size: pageSize,
    fields: `
      id
      title
      type
      state
      enabled
      type
      enabled
      start
      begin
      finish
      end
      comment
      description
      icon
      previewImage: preview_image
      priority
      ruleImage: rule_image
      state
      tags
      title
      joinRewards: join_rewards
      url
      canExitStates: can_exit_states
      canJoinStates: can_join_states
      joinTicket: join_ticket
      notices
    `,
  });
}

/**获取活动详情 */
export function getActivityInfo(request, id: number) {
  return extendQuery(request, {}, '/v1/activity')
    .query(
      `
    query infoActivity($id: ID!) {
      details: activities(id: $id) {
        id
        title
        type
        state
        enabled
        type
        enabled
        start
        begin
        finish
        end
        comment
        description
        icon
        previewImage: preview_image
        priority
        ruleImage: rule_image
        state
        tags
        title
        joinRewards: join_rewards
        url
        canExitStates: can_exit_states
        canJoinStates: can_join_states
        joinTicket: join_ticket
        notices
        class
        hookUrl: hook_url
        retainDuration: retain_duration
      }
    }
    `,
      {
        vars: {
          id,
        },
      },
    )
    .then((res) => res.details || {});
};

/**patch */
export function patchActivity(request, body) {
  return extendQuery(request, {})('/activity', {
    method: 'PATCH',
    data: body,
  });
}
