// mobx-sync/lib/sync.js
"use strict";
/*!
 *
 * Copyright 2017 - acrazing
 *
 * @author acrazing joking.young@gmail.com
 * @since 2017-11-28 17:31:44
 * @version 1.0.0
 * @desc sync.ts
 */
Object.defineProperty(exports, "__esModule", { value: true });
var mobx_1 = require("mobx");
var consts_1 = require("monofile-utilities/lib/consts");
var keys_1 = require("./keys");
var parse_store_1 = require("./parse-store");
var SyncTrunk = /** @class */ (function () {
    function SyncTrunk(store, _a) {
        var _b = _a === void 0 ? {} : _a, _c = _b.storage, storage = _c === void 0 ? localStorage : _c, _d = _b.storageKey, storageKey = _d === void 0 ? keys_1.KeyDefaultKey : _d, _e = _b.delay, delay = _e === void 0 ? 0 : _e, _f = _b.onError, onError = _f === void 0 ? consts_1.noop : _f;
        this.store = store;
        this.storage = storage;
        this.storageKey = storageKey;
        this.delay = delay;
        this.onError = onError;
    }
    SyncTrunk.prototype.persist = function () {
        try {
            this.storage.setItem(this.storageKey, JSON.stringify(mobx_1.toJS(this.store)));
        }
        catch (error) {
            this.onError(error);
        }
    };
    /**
     * init the store
     */
    SyncTrunk.prototype.init = function (initialState) {
        try {
            var data = this.storage.getItem(this.storageKey);
            if (data) {
                parse_store_1.parseStore(this.store, JSON.parse(data), false);
            }
        }
        catch (_a) {
            // DO nothing
        }
        if (initialState) {
            parse_store_1.parseStore(this.store, initialState, true);
        }
        // persist before listen change
        this.persist();
        this.disposer = mobx_1.autorun(this.persist.bind(this), {
            name: keys_1.KeyActionName,
            delay: this.delay,
            onError: this.onError,
        });
    };
    SyncTrunk.prototype.clear = function () {
        this.storage.removeItem(this.storageKey);
    };
    SyncTrunk.prototype.updateStore = function (store) {
        this.store = store;
        this.persist();
    };
    return SyncTrunk;
}());
exports.SyncTrunk = SyncTrunk;
